src/EditorFeatures/Core/Implementation/CodeActions/CodeActionEditHandlerService.cs

```diff
@@ -191,84 +191,88 @@ public ITextBufferAssociatedViewService AssociatedViewService
-
-                    // 0 file changes
-                    if (changedFiles.Count == 0)
-                    {
-                        operation.Apply(workspace, cancellationToken);
-                        continue;
-                    }
+            var applyOperation = operationsList.Single() as ApplyChangesOperation;
+            if (applyOperation == null)
+            {
+                return null;
+            }
 
-                    // 1 file change
-                    SourceText text = null;
-                    if (changedFiles.Count == 1)
-                    {
-                        if (changedDocuments.Any())
-                        {
-                            // ConfigureAwait(true) so we come back to the same thread as 
-                            // we do all application on the UI thread.
-                            text = await oldSolution.GetDocument(changedDocuments.Single()).GetTextAsync(cancellationToken).ConfigureAwait(true);
-                        }
-                        else if (changedAdditionalDocuments.Any())
-                        {
-                            // ConfigureAwait(true) so we come back to the same thread as 
-                            // we do all application on the UI thread.
-                            text = await oldSolution.GetAdditionalDocument(changedAdditionalDocuments.Single()).GetTextAsync(cancellationToken).ConfigureAwait(true);
-                        }
-                    }
+            var newSolution = applyOperation.ChangedSolution;
+            var changes = newSolution.GetChanges(oldSolution);
 
-                    if (text != null)
-                    {
-                        using (workspace.Services.GetService<ISourceTextUndoService>().RegisterUndoTransaction(text, title))
-                        {
-                            operation.Apply(workspace, cancellationToken);
-                            continue;
-                        }
-                    }
+            if (changes.GetAddedProjects().Any() ||
+                changes.GetRemovedProjects().Any())
+            {
+                return null;
+            }
 
-                    // multiple file changes
-                    using (var undoTransaction = workspace.OpenGlobalUndoTransaction(title))
-                    {
-                        operation.Apply(workspace, progressTracker, cancellationToken);
+            var projectChanges = changes.GetProjectChanges().ToImmutableArray();
+            if (projectChanges.Length != 1)
+            {
+                return null;
+            }
 
-                        // link current file in the global undo transaction
-                        if (fromDocument != null)
-                        {
-                            undoTransaction.AddDocument(fromDocument.Id);
-                        }
+            var projectChange = projectChanges.Single();
```

Comment:
If you use `SingleOrDefault()`, you could avoid the allocation of an `ImmutableArray<T>` and iterating the whole `IEnumerable<ProjectChanges>`.